### PR TITLE
Refresh instance metadata on boot

### PR DESCRIPTION
**Pull request by endgame**
###### Motivation for this change

Re-fetch instance meta data/user data on each boot. This means that we can stop instances, change user data, reboot the instance, and expect the new user data to actually be applied.

~~**NOTE** I tried to cargo-cult the docbook but I must have screwed it up. Help please?~~ Never mind, my test instance ran out of memory trying to rebuild docs.

###### Things done

* Built an EC2 AMI at commit f2ca527 (i.e., everything but the docbook changes), launched that AMI with no user data, stopped the instance, set up some user data and rebooted the instance, and checked that the user data was being applied.

Ping @grahamc

### CODE REVIEW

**Review by grahamc — COMMENTED** [review]
I suppose this doesn't change anything about how we support NixOps, right? If there was userdata for a nixops-managed host, this has the same semantics -- just it'd fetch newer metadata on boot?

**Comment by endgame** [comment]
> I suppose this doesn't change anything about how we support NixOps, right? If there was userdata for a nixops-managed host, this has the same semantics -- just it'd fetch newer metadata on boot?

Nixops will not change the ssh hostkeys even on userdata change, because the hostkey will still exists: https://github.com/NixOS/nixpkgs/blob/ae94e8923240f7ff7e82abf4783ef4318b8c4464/nixos/modules/virtualisation/ec2-data.nix#L51

I think it will still run the same amazon-init service and reapply configuration.nix userdata: https://github.com/NixOS/nixpkgs/blob/3bea94027167951474fbdf0569ce26d176260139/nixos/modules/virtualisation/amazon-init.nix#L19

Should any of this discussion be replicated in the release notes? Also, do you have any idea what might be going wrong with the docbook stuff?

File: nixos/doc/manual/release-notes/rl-2103.xml [hunk 1]
```diff
@@ -211,6 +211,11 @@
       and <literal>slaptest</literal> is buggy with schemas directly in the config file.
     </para>
    </listitem>
+   <listitem>
+     <para>
+       AWS EC2 and OpenStack compute images now re-fetch instance meta data and user data on each boot. For example: stopping an EC2 instance, changing its user data, and restarting the instance will cause it to fetch and apply the new user data instead of the old.
```
Comment (grahamc) [inline nixos/doc/manual/release-notes/rl-2103.xml:216]:
```suggestion
       Amazon EC2 and OpenStack compute images now re-fetch instance meta data and user data on each boot. For example: stopping an EC2 instance, changing its user data, and restarting the instance will cause it to fetch and apply the new user data.
```

(nit: product name is Amazon EC2)

Comment (endgame) [reply nixos/doc/manual/release-notes/rl-2103.xml]: Done

**Review by grahamc — APPROVED** [review]

**Pull request merged**